src/builder/modules/Generator_Builder.res

```diff
@@ -439,6 +439,67 @@ let fillBackgroundColor = (model: Model.t, fillStyle: string) => {
   }
 }
 
+let getOffset = ((x1, y1): position, (x2, y2): position) => {
+  let x1 = Belt.Int.toFloat(x1)
+  let y1 = Belt.Int.toFloat(y1)
+  let x2 = Belt.Int.toFloat(x2)
+  let y2 = Belt.Int.toFloat(y2)
+
+  let w = x2 -. x1
+  let h = y2 -. y1
+
+  /* When a line is drawn and its start and end coords are integer values, the
+  resulting line is drawn in between to rows of pixels, resulting in a line
+  that is two pixels wide and half transparent. To fix this, the line's start
+  and end positions need to be offset 0.5 pixels in the direction normal to the
+  line. The following code gets the angle of the line, and gets the components
+  for a translation in the direction perpendicular to the angle using vector
+  resolution: https://physics.info/vector-components/summary.shtml This results
+  in a fully opaque line with the correct width if the line is vertical or
+  horizontal, but antialiasing may still affect lines at other angles.
+ */
+  let angle = w === 0.0 ? Js.Math._PI /. 2.0 : Js.Math.atan2(~y=h, ~x=w, ())
```

Comment:
Just thinking about this - do we need the special case for vertical lines? Would this be fine?

```res
let angle = Js.Math.atan2(~y=h, ~x=w, ())
```

Reply:
The atan2() function takes the arctangent of the height divided by the width (the slope) to get the angle, but having a width of zero means dividing be zero, which gives an error. Because of this, the angle in radians for a 90 degree slope needs to be hard coded to pi / 2.

Reply:
I also realised that the direction the line is drawn (point A to B, vs point B to A) changes the direction of the normal, which then causes the offset to be either negative or positive - which then changes the position of the drawn line.

But I think it's fine, it's a limitation of having two different coordinate systems - one system for pixel coordinates, and another system for line point coordinates.

Reply:
> The atan2() function takes the arctangent of the height divided by the width (the slope) to get the angle, but having a width of zero means dividing be zero, which gives an error. Because of this, the angle in radians for a 90 degree slope needs to be hard coded to pi / 2.

I guessed something like that was the case in the original code, but maybe the JS `atan2()` function is handling it? Can you confirm?

<img width="170" alt="image" src="https://user-images.githubusercontent.com/85809598/170849176-75c7de0e-fa5b-4e1a-ad9b-1962120c9a55.png">

Reply:
It seems like it does work without hard coding- I think that when I first tested it, it didn't work because I was using the regular atan() function and dividing normally, which would have given an error.